### PR TITLE
512 uncertainty herding

### LINKED ISSUE #512: FR: Implement the pool-based query strategy `UncertaintyHerding`
### Describe the workflow you want to enable

`UncertaintyHerding` [1] is a pool-based query strategy suited for all budgets in active learning scenarios.

[1] [Bae, Wonho, Danica J. Sutherland, and Gabriel L. Oliveira. "Uncertainty Herding: One Active Learning Method for All Label Budgets." In ICLR. 2025.](https://openreview.net/forum?id=UgPoHhYQ2Uf)

### Describe your proposed solution

The authors describe `UncertaintyHerding` in Algorithm 1 of their article [1].

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mherde**
Closes #512

**Additional context**
> Implemented calibration as option, if the classifier supports outputting class logits. Classifiers outputting logits and embeddings via `predict_proba` or the ones implementing `decision_function` are allowed.

### CODE REVIEW

File: docs/examples/1-pool-classification/_uherding.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    "title" : "Uncertainty Herding (UHerding)",
+    "refs" : ["bae2025uncertainty"],
+    "sequence" : ["title", "text_0", "plot", "refs"],
+    "text_0" : "**Idea:** UHerding greedily selects a batch by maximizing uncertainty-weighted coverage in feature space. It combines predictive uncertainty from a classifier with a kernel-based coverage objective and adapts the temperature scaling and kernel radius from the currently labeled data. Here, the query-side classifier uses an RBF feature map with logistic regression so that uncertainty estimates remain expressive on the non-linearly separable pool data. Logits are preferred because they enable temperature scaling directly; if they are unavailable, UHerding falls back to the classifier's probability estimates.",
+    "import_clf" : ["from sklearn.kernel_approximation import RBFSampler",
+                    "from sklearn.linear_model import LogisticRegression",
+                    "from sklearn.pipeline import make_pipeline",
+                    "from skactiveml.classifier import ParzenWindowClassifier",
+                    "from skactiveml.classifier import SklearnClassifier"],
```
Comment (tpham93) [inline docs/examples/1-pool-classification/_uherding.json]: Do we really need all these imports? From what I can see, we only need the `ParzenWindowClassifier` import.

File: docs/examples/1-pool-classification/_uherding.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    "class" : "UHerding",
+    "package" : "pool",
+    "method" : "Uncertainty Herding (UHerding)",
+    "category" : "Hybrid",
+    "template" : "examples/template_pool.py",
+    "tags" : ["pool", "classification", "single-annotator", "diverse-batch"],
+    "title" : "Uncertainty Herding (UHerding)",
+    "refs" : ["bae2025uncertainty"],
+    "sequence" : ["title", "text_0", "plot", "refs"],
+    "text_0" : "**Idea:** UHerding greedily selects a batch by maximizing uncertainty-weighted coverage in feature space. It combines predictive uncertainty from a classifier with a kernel-based coverage objective and adapts the temperature scaling and kernel radius from the currently labeled data. Here, the query-side classifier uses an RBF feature map with logistic regression so that uncertainty estimates remain expressive on the non-linearly separable pool data. Logits are preferred because they enable temperature scaling directly; if they are unavailable, UHerding falls back to the classifier's probability estimates.",
```
Comment (mherde) [inline docs/examples/1-pool-classification/_uherding.json]:
```suggestion
    "text_0" : "**Idea:** UHerding greedily selects a batch by maximizing uncertainty-weighted coverage in feature space. It combines predictive uncertainty from a classifier with a kernel-based coverage objective and adapts the temperature scaling and kernel radius from the currently labeled data. Logits are preferred because they enable temperature scaling directly; if they are unavailable, UHerding falls back to the classifier's probability estimates.",
```

**Review by tpham93 — CHANGES_REQUESTED** [review]

File: docs/examples/1-pool-classification/_uherding.json (hunk 1 above)
Comment (mherde) [reply docs/examples/1-pool-classification/_uherding.json]: No

Comment (mherde) [inline docs/examples/1-pool-classification/_uherding.json]:
```suggestion
```

File: docs/generate.py [hunk 3]
```diff
@@ -22,11 +23,39 @@
 
 import skactiveml
 
+try:
+    import black
+except ModuleNotFoundError:
+    black = None
+
```
Comment (mherde) [inline docs/generate.py]:
```suggestion
```

File: docs/generate.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import os
 from collections import OrderedDict
+from pathlib import Path
```
Comment (mherde) [inline docs/generate.py:3]:
```suggestion
from pathlib import Path
import black
```

File: docs/generate.py [hunk 5]
```diff
@@ -22,11 +23,39 @@
...
+
+    if black is None:
+        if not _BLACK_WARNING_EMITTED:
+            warnings.warn(
+                "Black is not installed; generated example scripts will not "
+                "be reformatted."
+            )
+            _BLACK_WARNING_EMITTED = True
+        return
+
```
Comment (mherde) [inline docs/generate.py]:
```suggestion
def _format_generated_python_script(file_path, line_length=79):
    """Formats a generated Python example script with Black."""
```

File: skactiveml/pool/_uherding.py [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+    labeled_indices,
+    rand_argmax,
+)
+from ._uncertainty_sampling import uncertainty_scores
+
+
+class UHerding(SingleAnnotatorPoolQueryStrategy):
+    """Uncertainty Herding (UHerding)
+
+    "Uncertainty Herding" (UHering) is a query strategy [1]_ that
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
    "Uncertainty Herding" (UHerding) is a query strategy [1]_ that
```

File: skactiveml/pool/_uherding.py [hunk 7]
```diff
@@ -0,0 +1,566 @@
...
+    greedily maximizes an uncertainty-weighted coverage objective in feature
+    space. In addition to the greedy selection itself, the implementation
+    follows the paper-faithful parameter adaptation scheme:
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
    follows the parameter adaptation scheme of the paper:
```

File: skactiveml/pool/_uherding.py [hunk 8]
```diff
@@ -0,0 +1,566 @@
...
+        method="margin_sampling",
+        predict_proba_dict=None,
+        predict_proba_parser=None,
+        temperatures=None,
+        validation_size=0.2,
+        n_ece_bins=15,
+        normalize_samples=True,
+        adaptive_sigma=True,
+        metric="rbf",
+        metric_dict=None,
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
        metric="rbf",
        metric_dict=None,
        adaptive_sigma=True,
```

File: skactiveml/pool/_uherding.py [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+    normalize_samples : bool, default=True
+        Flag whether to normalize feature vectors to unit length before
+        computing pairwise distances and kernels.
+    adaptive_sigma : bool, default=True
+        Flag whether to adapt the radius according to the minimum non-zero
+        labeled pairwise distance. This option requires `metric='rbf'`.
+    metric : str or callable, default='rbf'
+        Kernel used for the coverage objective.
+    metric_dict : dict or None, default=None
+        Optional keyword arguments passed to `pairwise_kernels`.
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
    metric : str or callable, default='rbf'
        Kernel used for the coverage objective.
    metric_dict : dict or None, default=None
        Optional keyword arguments passed to `pairwise_kernels`.
    adaptive_sigma : bool, default=True
        Flag whether to adapt the radius according to the minimum non-zero
        labeled pairwise distance. This option requires `metric='rbf'`.
```

File: skactiveml/pool/_uherding.py [hunk 10]
```diff
@@ -0,0 +1,566 @@
...
+        self.method = method
+        self.predict_proba_dict = predict_proba_dict
+        self.predict_proba_parser = predict_proba_parser
+        self.temperatures = temperatures
+        self.validation_size = validation_size
+        self.n_ece_bins = n_ece_bins
+        self.normalize_samples = normalize_samples
+        self.adaptive_sigma = adaptive_sigma
+        self.metric = metric
+        self.metric_dict = metric_dict
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
        self.metric = metric
        self.metric_dict = metric_dict
        self.adaptive_sigma = adaptive_sigma
```

File: skactiveml/pool/_uherding.py [hunk 11]
```diff
@@ -0,0 +1,566 @@
...
+        sample_weight=None,
+        candidates=None,
+        batch_size=1,
+        return_utilities=False,
+    ):
+        """Determines for which candidate samples labels are to be queried.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
        X : array-like of shape (n_samples, ...)
```

File: skactiveml/pool/_uherding.py [hunk 12]
```diff
@@ -0,0 +1,566 @@
...
+            are used as fallback.
+        fit_clf : bool, default=True
+            Defines whether the classifier `clf` should be fitted on `X`, `y`,
+            and `sample_weight` before evaluating the acquisition function.
+            Independent of this flag, temporary cloned classifiers may still be
+            fitted internally to select the temperature parameter.
+        sample_weight : array-like of shape (n_samples,), default=None
+            Weights of training samples in `X`.
+        candidates : None or array-like of shape (n_candidates,), dtype=int \
+                or array-like of shape (n_candidates, n_features), default=None
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
                or array-like of shape (n_candidates, ...), default=None
```

File: skactiveml/pool/_uherding.py [hunk 13]
```diff
@@ -0,0 +1,566 @@
...
+            - If `candidates` is `None`, the unlabeled samples from `(X, y)`
+              are considered as candidates.
+            - If `candidates` is of shape `(n_candidates,)` and of type
+              `int`, `candidates` is considered as the indices of the samples
+              in `(X, y)`.
+            - If `candidates` is of shape `(n_candidates, n_features)`, the
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
            - If `candidates` is of shape `(n_candidates, ...)`, the
```

File: skactiveml/pool/_uherding.py [hunk 14]
```diff
@@ -0,0 +1,566 @@
...
+            e.g., `utilities[0]` indicates the utilities used for selecting
+            the first sample (with index `query_indices[0]`) of the batch.
+            Utilities for labeled samples or already selected candidates are
+            set to `np.nan`.
+
+            - If `candidates` is `None`, the indexing refers to the samples
+              in `X`.
+            - If `candidates` is of shape `(n_candidates,)` and of type
+              `int`, `utilities` refers to the samples in `X`.
+            - If `candidates` is of shape `(n_candidates, n_features)`,
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
            - If `candidates` is of shape `(n_candidates, ...)`,
```

File: skactiveml/pool/_uherding.py [hunk 15]
```diff
@@ -0,0 +1,566 @@
...
+        Returns
+        -------
+        query_indices : numpy.ndarray of shape (batch_size,)
+            The query indices indicate for which candidate sample a label is
+            to be queried, e.g., `query_indices[0]` indicates the first
+            selected sample.
+
+            - If `candidates` is `None` or of shape `(n_candidates,)`, the
+              indexing refers to the samples in `X`.
+            - If `candidates` is of shape `(n_candidates, n_features)`, the
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
            - If `candidates` is of shape `(n_candidates, ...)`, the
```

File: skactiveml/pool/_uherding.py [hunk 16]
```diff
@@ -0,0 +1,566 @@
...
+            indicator_funcs=[callable],
+        )
+        check_type(self.metric_dict, "metric_dict", (dict, type(None)))
+        metric_dict = (
+            {} if self.metric_dict is None else self.metric_dict.copy()
+        )
+        if self.adaptive_sigma and self.metric != "rbf":
+            raise ValueError(
+                "`adaptive_sigma=True` is only supported with `metric='rbf'`."
+            )
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
    if self.adaptive_sigma:
        if self.metric != "rbf":
            raise ValueError(
                "`adaptive_sigma=True` is only supported with `metric='rbf'`."
            )
        elif "gamma" in metric_dict:
            raise ValueError(
                "`'gamma' cannot be part of the `metric_dict` "
                "with `adaptive_sigma=True`."
            )
```

File: skactiveml/pool/_uherding.py [hunk 17]
```diff
@@ -0,0 +1,566 @@
...
+        # (Re-)fit classifier on full labeled data if requested.
+        if fit_clf:
+            if sample_weight is None:
+                clf_eval = clone(clf).fit(X, y)
+            else:
+                clf_eval = clone(clf).fit(X, y, sample_weight)
+        else:
+            clf_eval = clf
+
+        # Infere probabilities and if available logits as well as embeddings.
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
        # Infer probabilities and if available logits as well as embeddings.
```

File: skactiveml/pool/_uherding.py [hunk 18]
```diff
@@ -0,0 +1,566 @@
...
+        probas_cand, logits_cand, X_cand_repr = self._predict_with_extras(
+            clf_eval, X_cand
+        )
+        if X_cand_repr is None:
+            X_cand_repr = X_cand
+        if logits_cand is not None:
+            probas_cand = softmax(logits_cand / tau, axis=1)
+
+        # Compute uncertatiny scores by either using the original probability
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
        # Compute uncertainty scores by either using the original probability
```

File: skactiveml/pool/_uherding.py [hunk 19]
```diff
@@ -0,0 +1,566 @@
...
+                X_labeled_repr = X[labeled_idx]
+
+        # Normalize candidate and labeled samples to unit length.
+        if self.normalize_samples:
+            X_cand_repr = normalize(X_cand_repr, copy=True)
+            if X_labeled_repr is not None:
+                X_labeled_repr = normalize(X_labeled_repr, copy=True)
+
+        # Compute kernel similarities, where the bandwidth is automatically
+        # tune if an RBF kernel is employed.
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
        # tuned if an RBF kernel is employed.
```

File: skactiveml/pool/_uherding.py [hunk 20]
```diff
@@ -0,0 +1,566 @@
...
+        return query_indices
+
+    def _select_temperature(self, X, y, clf, temperatures, sample_weight=None):
+        # Fallback if there is only one temperature candidate.
+        if np.isscalar(temperatures):
+            return float(temperatures)
+        if len(temperatures) == 1:
+            return float(temperatures[0])
+
+        # Try to perform train-test split. If it not possilbe, return 1.0 as
```
Comment (mherde) [inline skactiveml/pool/_uherding.py:377]:
```suggestion
        # Try to perform train-test split. If not possible, return 1.0 as
```

File: skactiveml/pool/_uherding.py [hunk 21]
```diff
@@ -0,0 +1,566 @@
...
+            if self.predict_proba_dict is None
+            else self.predict_proba_dict.copy()
+        )
+        try:
+            out = clf.predict_proba(X, **predict_proba_dict)
+        except TypeError:
+            if predict_proba_dict:
+                out = clf.predict_proba(X)
+            else:
+                raise
```
Comment (mherde) [inline skactiveml/pool/_uherding.py]:
```suggestion
        out = clf.predict_proba(X, **predict_proba_dict)
```

**Review by tpham93 — APPROVED** [review]
